### PR TITLE
add custom text logger

### PR DESCRIPTION
**Pull request by n01z3**

### CODE REVIEW

File: dl/callbacks.py [hunk 1]
```diff
@@ -174,39 +185,96 @@ def on_batch_end(self, state):
 
         self.time = time.time()
         state.step += bs
-        state.batch_metrics = defaultdict(lambda: 0)
+        #
```
Comment (Scitator) [inline dl/callbacks.py]: to remove?

File: dl/callbacks.py [hunk 2]
```diff
@@ -174,39 +185,96 @@ def on_batch_end(self, state):
...
         state.epoch_metrics[lm] = {
             key: UtilsFactory.get_val_from_metric(value)
-            for key, value in state.epoch_metrics[lm].items()}
+            for key, value in state.epoch_metrics[lm].items()
+        }
 
         for key, value in state.epoch_metrics[lm].items():
             self.loggers[lm].add_scalar(f"epoch {key}", value, state.epoch)
 
         epoch_metrics_str = "\t".join([
```
Comment (Scitator) [inline dl/callbacks.py]: place under `if self.verbose:`

File: dl/callbacks.py [hunk 3]
```diff
@@ -174,39 +185,96 @@ def on_batch_end(self, state):
...
+        if self.verbose:
+            print("{epoch} * Epoch ({mode}): {metrics}".format(
+                epoch=state.epoch, mode=lm, metrics=epoch_metrics_str))
 
         if self.reset_step:
             state.step = None
 
 
+class Logger(Callback):
+    def __init__(self, log_dir):
```
Comment (Scitator) [inline dl/callbacks.py]: `logdir` and annotations please

File: dl/callbacks.py [hunk 4]
```diff
@@ -174,39 +185,96 @@ def on_batch_end(self, state):
...
+        super().__init__()
+        os.makedirs(log_dir, exist_ok=True)
+        log_filepath = os.path.join(log_dir, 'n01_logs.txt')
```
Comment (Scitator) [inline dl/callbacks.py]: txt_logs.txt?

File: dl/callbacks.py [hunk 5]
```diff
@@ -174,39 +185,96 @@ def on_batch_end(self, state):
...
+        self.logger = self._get_logger(log_filepath)
+
+    def on_epoch_begin(self, state):
+        self.logger.info('Epoch {} | optimizer "{}" | lr {}'.format(
+            state.epoch, state.optimizer.__class__.__name__,
+            self._get_current_lr(state)))
+
+    def on_epoch_end(self, state):
+        self.logger.info("Train metrics: " + self._get_metrics_string(
```
Comment (Scitator) [inline dl/callbacks.py]: goto
```
for key, value in state.epoch_metrics.items()
    ...
```
?

File: dl/callbacks.py [hunk 6]
```diff
@@ -174,39 +185,96 @@ def on_batch_end(self, state):
...
+        self.logger.info("Valid metrics: " + self._get_metrics_string(
+            state.epoch_metrics['valid']) + "\n")
+
+    def on_train_begin(self, state):
+        self.logger.info(
+            'Starting training with params:\n{}\n\n'.format(state))
+
+    def on_batch_end(self, state):
+        if state.is_train:
+            # print(self.state.epoch_metrics['train'])
```
Comment (Scitator) [inline dl/callbacks.py:241]: to remove

File: dl/callbacks.py [hunk 7]
```diff
@@ -174,39 +185,96 @@ def on_batch_end(self, state):
...
+            state.epoch_metrics['train']))
+        self.logger.info("Valid metrics: " + self._get_metrics_string(
+            state.epoch_metrics['valid']) + "\n")
+
+    def on_train_begin(self, state):
+        self.logger.info(
+            'Starting training with params:\n{}\n\n'.format(state))
+
+    def on_batch_end(self, state):
+        if state.is_train:
```
Comment (Scitator) [inline dl/callbacks.py:240]: why only `is_train`?

File: dl/callbacks.py [hunk 8]
```diff
@@ -174,39 +185,96 @@ def on_batch_end(self, state):
...
+        formatter = logging.Formatter('[%(asctime)s] %(message)s')
+        fh.setFormatter(formatter)
+        ch.setFormatter(formatter)
+        # add the handlers to the logger
+        logger.addHandler(fh)
+        logger.addHandler(ch)
+        return logger
+
+    def _get_current_lr(self, state):
+        lr_cur = state.optimizer['main'].param_groups[0]['lr']
```
Comment (Scitator) [inline dl/callbacks.py]: use `state.lr` instead?

File: dl/callbacks.py [hunk 9]
```diff
@@ -301,10 +372,11 @@ def pack_checkpoint(self, **kwargs):
         return UtilsFactory.pack_checkpoint(**kwargs)
 
     @staticmethod
-    def process_epoch_metrics(
-            epoch_metrics, best_metrics,
-            valid_loader="valid",
-            main_metric="loss", minimize=True):
+    def process_epoch_metrics(epoch_metrics,
```
Comment (Scitator) [inline dl/callbacks.py:374]: formatting

File: dl/runner.py [hunk 10]
```diff
@@ -65,11 +65,8 @@ def _init(self):
         """
         self.model, self.device = UtilsFactory.prepare_model(self.model)
 
-    def _init_state(
-            self, *,
-            mode: str,
-            stage: str = None,
-            **kwargs) -> RunnerState:
+    def _init_state(self, *, mode: str, stage: str = None,
```
Comment (Scitator) [inline dl/runner.py:68]: formatting

File: dl/state.py [hunk 11]
```diff
@@ -14,6 +14,7 @@ def __init__(self, **kwargs):
         self.logdir = None
         self.loader = None
         self.loader_mode = None
+        self.batch_size = 8
```
Comment (Scitator) [inline dl/state.py:17]: fix

File: dl/callbacks.py (hunk 4 above)
Comment (n01z3) [reply dl/callbacks.py]: yes, it's better

File: dl/callbacks.py (hunk 7 above)
Comment (n01z3) [reply dl/callbacks.py:240]: Because in validation mode i'm intersected only in total metrics

**Pull request closed without merging**